File: solutions/java/say/2/src/main/java/Say.java
import java.util.Map;
import java.util.Objects;
import java.util.stream.Collectors;
import java.util.stream.IntStream;

public class Say {
	private static final String[] BELOW_TWENTY = {
			"zero", "one", "two", "three", "four", "five",
			"six", "seven", "eight", "nine", "ten",
			"eleven", "twelve", "thirteen", "fourteen", "fifteen",
			"sixteen", "seventeen", "eighteen", "nineteen"
	};

	private static final Map<Integer, String> SINGLE_DIGIT_CHUNKS
			= IntStream.range(1, 20)
                        .boxed()
                        .collect(Collectors.toMap(i -> i, i -> BELOW_TWENTY[i]));

	private static final Map<Integer, String> TENS = Map.ofEntries(
			Map.entry(20, "twenty"),
			Map.entry(30, "thirty"),
			Map.entry(40, "forty"),
			Map.entry(50, "fifty"),
			Map.entry(60, "sixty"),
			Map.entry(70, "seventy"),
			Map.entry(80, "eighty"),
			Map.entry(90, "ninety")
	);

	private static final String[] PERIOD_NAMES = {"", "thousand", "million", "billion"};

	public String say(long number) {
		if (number < 0 || number > 999_999_999_999L) {
			throw new IllegalArgumentException("Number out of range");
		}

		return this.streamOverThousandsChunks(number);
	}

	private String streamOverThousandsChunks(long number) {
		return IntStream.range(0, PERIOD_NAMES.length)
				       .mapToObj(i -> chunkToWords(number, i))
				       .filter(Objects::nonNull)
				       .reduce(this::toProperWordOrder)
				       .orElse("zero");
	}

	private String chunkToWords(long number, int periodIndex) {
		double periodBase = Math.pow(1000, periodIndex);
		double periodValue = number / periodBase;
		int chunk = (int) (periodValue % 1000);
		return chunk == 0
				       ? null
				       : this.toWords(chunk, periodIndex);

	}

	private String toWords(int chunk, int periodIndex) {
		String chunkWords = this.convertChunk(chunk);
		String periodLabel = PERIOD_NAMES[periodIndex].isEmpty()
				                     ? ""
				                     : " " + PERIOD_NAMES[periodIndex];
		return chunkWords + periodLabel;
	}

	private String toProperWordOrder(String a, String b) {
		return b + " " + a;
	}

	private String convertChunk(int number) {
		return number >= 100
				       ? this.convertOver100Inclusive(number)
				       : number >= 20
						         ? this.convertOver20Inclusive(number)
						         : SINGLE_DIGIT_CHUNKS.get(number);
	}

	private String convertOver100Inclusive(int number) {
		int hundreds = number / 100;
		int remainder = number % 100;
		String remainderWords = remainder != 0
				                        ? " " + this.convertChunk(remainder)
				                        : "";
		return BELOW_TWENTY[hundreds] + " hundred" + remainderWords;
	}

	private String convertOver20Inclusive(int number) {
		int tensPart = (number / 10) * 10;
		int units = number % 10;
		String unitsWords = units != 0
				                    ? "-" + BELOW_TWENTY[units]
				                    : "";
		return TENS.get(tensPart) + unitsWords;
	}
}

File: solutions/scala/say/2/src/main/scala/Say.scala
import scala.annotation.tailrec

object Say:
    private val ones: Map[Int, String] = Map(
        0 -> "zero",
        1 -> "one",
        2 -> "two",
        3 -> "three",
        4 -> "four",
        5 -> "five",
        6 -> "six",
        7 -> "seven",
        8 -> "eight",
        9 -> "nine",
        10 -> "ten",
        11 -> "eleven",
        12 -> "twelve",
        13 -> "thirteen",
        14 -> "fourteen",
        15 -> "fifteen",
        16 -> "sixteen",
        17 -> "seventeen",
        18 -> "eighteen",
        19 -> "nineteen"
    )

    private val tens: Map[Int, String] = Map(
        20 -> "twenty",
        30 -> "thirty",
        40 -> "forty",
        50 -> "fifty",
        60 -> "sixty",
        70 -> "seventy",
        80 -> "eighty",
        90 -> "ninety"
    )

    private val scales: Vector[String] = Vector(
        "",
        "thousand",
        "million",
        "billion",
        "trillion"
    )

    private def scaleWord(scaleIndex: Int): String =
        scales.lift(scaleIndex)
              .getOrElse("")

    def inEnglish(number: Long): Option[String] =
        if number < 0 || number > 999999999999L
        then None
        else Some(convertNumberToWords(number))

    private def convertNumberToWords(number: Long): String =
        if number == 0
        then "zero"
        else buildFullNumberWords(number)

    private def buildFullNumberWords(number: Long): String =
        val chunks = splitIntoThousands(number)
        val totalChunks = chunks.length
        chunks.zipWithIndex
              .map { case (chunk, index) =>
                  formatChunkWithScale(totalChunks, chunk, index)
              }
              .filter(_.nonEmpty)
              .mkString(" ")

    private def formatChunkWithScale(
                                        totalChunks: Int,
                                        chunk: Int,
                                        index: Int
                                    ): String =
        val scaleIndex = totalChunks - index - 1
        val chunkWords = convertBelowThousand(chunk)

        if chunkWords.isEmpty
        then ""
        else attachScale(scaleIndex, chunkWords)

    private def attachScale(scaleIndex: Int, words: String): String =
        val scale = scaleWord(scaleIndex)
        if scale.isEmpty
        then words
        else s"$words $scale"

    private def splitIntoThousands(number: Long): List[Int] =
        @tailrec
        def loop(x: Long, accumulator: List[Int]): List[Int] =
            if x == 0
            then accumulator
            else loop(x / 1000, (x % 1000).toInt :: accumulator)

        loop(number, Nil)

    private def convertBelowThousand(number: Int): String =
        number match
            case 0 => ""

            case x if x < 20 =>
                ones(x)

            case x if x < 100 =>
                val tensPart = (x / 10) * 10
                val onesPart = x % 10

                if onesPart == 0
                then tens(tensPart)
                else s"${tens(tensPart)}-${ones(onesPart)}"

            case x =>
                val hundreds = x / 100
                val remainder = x % 100

                if remainder == 0
                then s"${ones(hundreds)} hundred"
                else s"${ones(hundreds)} hundred ${convertBelowThousand(remainder)}"
                
                